src/EditorFeatures/CSharpTest/ConvertForToForEach/ConvertForToForEachTests.cs

```diff
@@ -0,0 +1,1268 @@
+@"using System;
+
+class C
+{
+    void Test(string[] array)
+    {
+        [||]for (int i = 0; i < array.Length; i++)
+        {
+            Console.WriteLine(array[i]);
+            Console.WriteLine(array[i]);
+        }
+    }
+}",
+@"using System;
+
+class C
+{
+    void Test(string[] array)
+    {
+        foreach (string {|Rename:v|} in array)
+        {
+            Console.WriteLine(v);
+            Console.WriteLine(v);
+        }
+    }
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsConvertForToForEach)]
+        public async Task TestEmbeddedStatement()
+        {
+            await TestInRegularAndScript1Async(
+@"using System;
+
+class C
+{
+    void Test(string[] array)
+    {
+        [||]for (int i = 0; i < array.Length; i++)
+            Console.WriteLine(array[i]);
+    }
+}",
+@"using System;
+
+class C
+{
+    void Test(string[] array)
+    {
+        foreach (string {|Rename:v|} in array)
+            Console.WriteLine(v);
+    }
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsConvertForToForEach)]
+        public async Task TestPostIncrement()
+        {
+            await TestInRegularAndScript1Async(
+@"using System;
+
+class C
+{
+    void Test(string[] array)
+    {
+        [||]for (int i = 0; i < array.Length; ++i)
+        {
+            Console.WriteLine(array[i]);
+        }
+    }
+}",
+@"using System;
+
+class C
+{
+    void Test(string[] array)
+    {
+        foreach (string {|Rename:v|} in array)
+        {
+            Console.WriteLine(v);
+        }
+    }
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsConvertForToForEach)]
+        public async Task TestArrayPlusEqualsIncrementor()
+        {
+            await TestInRegularAndScript1Async(
+@"using System;
+
+class C
+{
+    void Test(string[] array)
+    {
+        [||]for (int i = 0; i < array.Length; i += 1)
+        {
+            Console.WriteLine(array[i]);
+        }
+    }
+}",
+@"using System;
+
+class C
+{
+    void Test(string[] array)
+    {
+        foreach (string {|Rename:v|} in array)
+        {
+            Console.WriteLine(v);
+        }
+    }
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsConvertForToForEach)]
+        public async Task TestMissingOutsideKeyword()
+        {
+            await TestMissingInRegularAndScriptAsync(
+@"using System;
+
+class C
+{
+    void Test(string[] array)
+    {
+       [||] for (int i = 0; i < array.Length; i++)
+        {
+            Console.WriteLine(array[i]);
+        }
+    }
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsConvertForToForEach)]
+        public async Task TestMissingAfterOpenParen()
+        {
+            await TestMissingInRegularAndScriptAsync(
+@"using System;
+
+class C
+{
+    void Test(string[] array)
+    {
+        for ([||] int i = 0; i < array.Length; i++)
```

Comment:
@dotnet/roslyn-ide @jinujoseph @kuhlenh we probably need to decide where we should offer refactoring.

I am +1 on making it same as regular code fix unless refactoring require selection such as extract method.

basically working based on roughly line based.

Reply:
right now, different refactoring used different span to offer refactoring.

Reply:
> @dotnet/roslyn-ide @jinujoseph @kuhlenh we probably need to decide where we should offer refactoring.

This fits one of the patterns we've used for many (though not all) of our refactorings.  Namely that hte cursor should be on something that unambiguously represents the thing to be operated on.  So, for example, if you're refactoring a 'switch' statement, then you can offer on the switch keyword, or hte case keywords.  Similarly, for a 'for'-statement, we should not offer on sub-node of the statement, as those can be unbound components that the user would not think should affect the outer  for-loop when they touch.

Note: we updated many refactorings to follow this style due to having so many of them appear inappropriately.

--

Note: one thing we can consider is that *if* the construct is preceded by whitespace to the start of **that line**, we coudl also offer in that whitespace section.  However, we should not generally offer refactorings inside the leading trivia for a node that exists on *previous* lines.  These previous-line whitespace/comments are not what users think of as the place they woudl want to go to affect the statement on the next line.

Reply:
> basically working based on roughly line based.

Line based is usually quite inappropriate for a refactoring.  A line can contain a huge amount of different nodes and concepts.  For example, if i go into "va$$r x" and try to bring up refactoring to convert 'var' to 'int', then i do not want to see 'convert for to foreach'.  It's just clutter at best, and at worst, it will be prioritized poorly and it will impede the user from making the change.

Reply:
theoretically sure, it can have many. but we still show all fixes for diagnostics on same line. they do sort them, but I highly doubt whether there are many people who actually understand the order.

I think we choose line since it is well balanced between convenience vs too much offering.

I dont want to go specific "token" to get my refactoring or remember what the specific token is for the refactoring I want. if I choose to do "ctrl+.", most of time, it should offer what I want if I moved my caret closely enough.

unless we actually have a case where we show too many refactoring on a line that makes navigating offerings are more work than moving caret to exact position, I think following roughly same pattern as code fix is less burden to users.

Reply:
> theoretically sure, it can have many. but we still show all fixes for diagnostics on same line. they do sort them

Right.  But htat's because diagnostics have locations.  And thus it's fairly intuitive that the squiggle that is *closer* to your cursor is more relevant than one farther away.  No such location/sorting exists for refactorings.  A refactoring is just based on your cursor/selection.  So all refactorings are equally suitable.

> I think we choose line since it is well balanced between convenience vs too much offering.

We never chose 'line'.  And, indeed, we've take a ton of refactorings and made them non-line because they were showing up too often :)

> I dont want to go specific "token" to get my refactoring or remember what the specific token is for the refactoring I want. if I choose to do "ctrl+.", most of time, it should offer what I want if I moved my caret closely enough.

Yes.  That's what this PR does.  it keeps the feature available on intuitively understand parts of the for-statement, while ensuring it doesn't show up in areas i don't want it to (like when i'm on the 'var' part of a for-loop).  It's designed based on everything we've learned so far about refactorings showing up too much, and i'd like it to use those lessons and be consistent with them.

> unless we actually have a case where we show too many refactoring on a line that makes navigating offerings are more work than moving caret to exact position, I think following roughly same pattern as code fix is less burden to users.

This was not the case.  We found that having too many refactorings showing up *was* a burden to users.  So we spent a *lot* of time (seriously, tons and tons of fixes) to try to address this.  I do not want us rolling back here.

--

Note: as i mentioned above, i would be willing to offer this before the 'for' keyword, up to the start of the line it is on.  but that's the only extension to this check that would be reasonable and would not cause the same sort of problems we spent so much effort trying to address.

Reply:
whatever it is, I think all refactoring should follow roughly same pattern so that users can predict or have common expectation.

I expressed my opinion. so I guess need to hear other people's opinion. @jinujoseph @kuhlenh ?

my experience of refactoring is that I never goes to specific token or region, I use arrow key to move caret to a line and hit +. and search for refactoring I want. so caret's column will be whatever position it was before I move up and down using arrow key. so if its column was on right position, that is 100% by luck. not by intention.

Reply:
> whatever it is, I think all refactoring should follow roughly same pattern so that users can predict or have common expectation.

Agreed.  And that's why i am following the patterns we a couple of years ago as to where refactorings should be offered :)